File: SecondYearCS/OOP/Civitas/src/civitas/JugadorEspeculador.java
package civitas;

import java.util.ArrayList;
import java.util.Scanner;
import javax.swing.JOptionPane;


public class JugadorEspeculador extends Jugador {

    private int FactorEspeculador = 2;
    private Scanner in;
    private float fianza;
    
  
    
    public JugadorEspeculador(Jugador otro,float fianza) {
        super(otro);
        this.fianza = fianza;
        in = new Scanner (System.in);
    }
        
    @Override
     boolean pagaImpuesto(float cantidad){
        boolean result = false;
        if(!encarcelado){
            result = this.paga(cantidad/FactorEspeculador);
        }
        return result;
    }
     
     @Override
     protected int getCasasMax(){
         return super.getCasasMax()*FactorEspeculador;
     }
     
     @Override
     protected int getHotelesMax(){
         return super.getHotelesMax()*FactorEspeculador;
     }
    
      /**
     * @brief verifica si el jugador debe ser Encarcelado o No, y si posee SalvoConducto
     * @return verdadero/falso
     */
    @Override
    protected boolean debeSerEncarcelado(){
        boolean result = false;
        int numero = -1;
        if(encarcelado){
            result = true;
        }
        if(this.tieneSalvoconducto()){
            result = false;
            super.perderSalvoconducto();
            Diario.getInstance().ocurreEvento("\n\tSe Libre de la Cárcel\n\t");
        }else{
            //Ahora que estamos utilizando GUI
        numero = JOptionPane.showConfirmDialog(null, "¿Quieres comprar la calle actual?", "Compra", JOptionPane.YES_NO_OPTION);
        }
        if(super.puedoGastar(fianza) && numero == 0 ){
                result = false;
                super.paga(fianza);
                Diario.getInstance().ocurreEvento("\n\tNo Ha ido a Carcel, ya que ha pagado la fianza\n\t");
        }
        return result;
    }
}

File: SecondYearCS/OOP/CivitasRB/lib/JugadorEspeculador.rb
module CivitasRB
  class JugadorEspeculador < Jugador
    def initialize(otro,fianza,factorEspeculador=2)
      super(otro)
      @FactorEspeculador = factorEspeculador
      @fianza = fianza
    end

    def pagaImpuesto(cantidad)
      result = false
      if(!super.encarcelado)
        result = self.paga(cantidad/@FactorEspeculador)
      end
      return result
    end

    def getCasasMax()
      return super.getCasasMax()*@FactorEspeculador
    end

    def getHotelesMax()
      return super.getHotelesMax()*@FactorEspeculador
    end

    def debeSerEncarcelado
      result = false
      if(super.encarcelado)
        result = true
      end
      if(super.tieneSalvoconducto())
        result = false
        super.perderSalvoconducto()
        Diario.instance.ocurre_evento("\t\nSe Libra de la Carcel\t\n")
      else
        puts "Quieres pagar la fianza?\n\t0-Si\n\t1-No"
        loop do
           numero = gets.to_i
           if numero >=0 && numero < 2
              break
          end
        end
        if(super.puedoGastar(@fianza) && numero==0)
          result = false
          super.paga(fianza)
          Diario.instance.ocurre_evento("\n\tNo Ha ido a Carcel, ya que ha pagado la fianza\n\t")
        end
      end
      return result
    end

  end
end
